File: live_server/web/src/protocol.ts
export type RenderMode = "sequential" | "parallel";

export interface PixelWire {
  x: number;
  y: number;
  r: number;
  g: number;
  b: number; // u8 0..255
}

export type ServerMessage =
  | { type: "FrameStart"; width: number; height: number }
  | { type: "Pixels"; pixels: PixelWire[] }
  | { type: "FrameDone" };

export interface ClientStart {
  type: "Start";
  mode: RenderMode;
  batch_size?: number;
  scene?: string;
  width?: number;
  height?: number;
}

export interface ScenesResponse {
  scenes: string[];
}

export interface ResolutionPreset {
  label: string;
  width: number;
  height: number;
}

export interface ResolutionsResponse {
  resolutions: ResolutionPreset[];
}

File: live_server/src/protocol.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ServerMessage {
    FrameStart { width: usize, height: usize },
    Pixels { pixels: Vec<PixelWire> },
    FrameDone,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PixelWire {
    pub x: usize,
    pub y: usize,
    pub r: u8,
    pub g: u8,
    pub b: u8,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ClientMessage {
    Start {
        mode: RenderModeWire,
        #[serde(default)]
        batch_size: Option<usize>,
        #[serde(default)]
        scene: Option<String>,
        #[serde(default)]
        width: Option<usize>,
        #[serde(default)]
        height: Option<usize>,
    },
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum RenderModeWire {
    Sequential,
    Parallel,
}
